services/meta/data.go

```diff
@@ -447,6 +450,26 @@ func (data *Data) DeleteShardGroup(database, policy string, id uint64) error {
 	return ErrShardGroupNotFound
 }
 
+// PruneShardGroups removes any shards deleted before expiration and that have no remaining owners.
+// Returns true if data is modified.
+func (data *Data) PruneShardGroups(expiration time.Time) bool {
+	var changed bool
+	for i, d := range data.Databases {
+		for j, rp := range d.RetentionPolicies {
+			var remainingShardGroups []ShardGroupInfo
+			for _, sgi := range rp.ShardGroups {
+				if sgi.DeletedAt.IsZero() || !expiration.After(sgi.DeletedAt) || len(sgi.Shards) > 0 {
```

Comment:
And here we check if any shards are left in a shard group, and retain it if so.

Reply:
This keeps the shards we failed to remove from disk in the meta-data for a later attempt at deletion.

Reply:
This is the actual fix for the problem. Do not prune shard groups which have shards that could not be deleted.

Reply:
Yes, this is the heart of the fix that prevents orphaned TSM files that InfluxDB will not attempt to delete.